colossalai/inference/kvcache_manager.py

```diff
@@ -0,0 +1,92 @@
+
+    def _init_mem_states(self, size, dev):
+        self.mem_state = torch.ones((size,), dtype=torch.bool, device=dev)
+        self._mem_cum_sum = torch.empty((size,), dtype=torch.int32, device=dev)
+        self.indexes = torch.arange(0, size, dtype=torch.long, device=dev)
+
+    def _init_kv_buffers(self, size, dev, dtype, head_num, head_dim, layer_num):
+        self.key_buffer = [torch.empty((size, head_num, head_dim), dtype=dtype, device=dev) for _ in range(layer_num)]
+        self.value_buffer = [torch.empty((size, head_num, head_dim), dtype=dtype, device=dev) for _ in range(layer_num)]
+
+    @torch.no_grad()
+    def alloc(self, required_size):
+        if required_size > self.available_size:
+            self.logger.warning(f"warn no enough cache required_size {required_size} left_size {self.available_size}")
+            return None
+
+        torch.cumsum(self.mem_state, dim=0, dtype=torch.int32, out=self._mem_cum_sum)
+        select_index = torch.logical_and(self._mem_cum_sum <= required_size, self.mem_state == 1)
+        select_index = self.indexes[select_index]
+        self.mem_state[select_index] = 0
+        self.available_size -= len(select_index)
+        return select_index
+
+    @torch.no_grad()
+    def alloc_contiguous(self, required_size):
+        if required_size > self.available_size:
+            self.logger.warning(f"warn no enough cache required_size {required_size} left_size {self.available_size}")
+            return None
+
+        torch.cumsum(self.mem_state, dim=0, dtype=torch.int32, out=self._mem_cum_sum)
+        sum_size = len(self._mem_cum_sum)
+        loc_sums = self._mem_cum_sum[required_size - 1:] - self._mem_cum_sum[0:sum_size - required_size +
+                                                                             1] + self.mem_state[0:sum_size -
+                                                                                                 required_size + 1]
+        can_used_loc = self.indexes[0:sum_size - required_size + 1][loc_sums == required_size]
+        if can_used_loc.shape[0] == 0:
+            self.logger.info(
+                f"Not enough contiguous cache: required_size {required_size} left_size {self.available_size}")
+            return None
+        start_loc = can_used_loc[0]
+        select_index = self.indexes[start_loc:start_loc + required_size]
+
+        self.mem_state[select_index] = 0
+        self.available_size -= len(select_index)
+        start = start_loc.item()
+        end = start + required_size
+        return select_index, start, end
+
+    @torch.no_grad()
+    def free(self, free_index):
+        self.available_size += free_index.shape[0]
+        self.mem_state[free_index] = 1
+
+    @torch.no_grad()
+    def free_all(self):
+        self.available_size = len(self.mem_state)
+        self.mem_state[:] = 1
+        self.past_key_values_length = 0
+        self.logger.info("freed all space of memory manager")
```

Comment:
maybe add some basic tests for the `infer_state`

Reply:
will add some tests for `InferenceState` and kvcache manager

Reply:
Tests for kvcache manager added. Batch inference state will be tested with llama/bloom inference flow.